File: week3-divide/uniform.py
import sys, math

def read_input():
    input_data = sys.stdin.read().strip().split('\n')
    i = 0
    test_cases = []
    
    while i < len(input_data):
        n = int(input_data[i].strip())
        if n == 0:
            break
        i += 1
        
        points = []
        for _ in range(n):
            x, y = map(float, input_data[i].strip().split())
            points.append((x, y))
            i += 1
        
        test_cases.append(points)
    
    return test_cases

def brute_force(px):
    minimum = float('inf')
    result = None
    for i in range(len(px)):
        for j in range(i+1, len(px)):
            dist = (px[i][0] - px[j][0]) ** 2 + (px[i][1] - px[j][1]) ** 2
            if dist < minimum:
                minimum = dist
                result = (px[i], px[j])
    return result, math.sqrt(minimum)

def split_list(a_list):
    half = len(a_list)//2
    return a_list[:half], a_list[half:]

def closest_pair(px, py):
    if len(px) <= 3:
        return brute_force(px)
    
    Qx, Rx = split_list(px)
    xmax = Qx[-1][0]

    Qy = [p for p in py if p[0] <= xmax]
    Ry = [p for p in py if p[0] > xmax]

    (q0, q1), dist_q = closest_pair(Qx, Qy)
    (r0, r1), dist_r = closest_pair(Rx, Ry)
    
    ss = min(dist_q, dist_r)
    best_pair = (q0, q1) if dist_q < dist_r else (r0, r1)

    Sy = [p for p in py if abs(p[0] - xmax) < ss]

    for i in range(len(Sy)):
        for j in range(i+1, min(i+7, len(Sy))):
            dist_sq = (Sy[i][0] - Sy[j][0]) ** 2 + (Sy[i][1] - Sy[j][1]) ** 2
            if dist_sq < ss ** 2:
                ss = math.sqrt(dist_sq)
                best_pair = (Sy[i], Sy[j])

    return best_pair, ss


test_cases = read_input()

for points in test_cases:
    
    px = sorted(points, key=lambda x: x[0])
    py = sorted(points, key=lambda x: x[1])
    
    res, dist = closest_pair(px, py)
    
    print(f'{res[0][0]} {res[0][1]} {res[1][0]} {res[1][1]}')

File: week3-divide/uniform.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Point struct {
	x, y float64
}

func Distance(p1, p2 Point) float64 {
	return math.Hypot(p1.x-p2.x, p1.y-p2.y)
}

func BruteForce(points []Point) (Point, Point, float64) {
	minDist := math.MaxFloat64
	var p1, p2 Point
	for i := 0; i < len(points); i++ {
		for j := i + 1; j < len(points); j++ {
			dist := Distance(points[i], points[j])
			if dist < minDist {
				minDist = dist
				p1 = points[i]
				p2 = points[j]
			}
		}
	}
	return p1, p2, minDist
}

func ClosestPair(px, py []Point) (Point, Point, float64) {
	if len(px) <= 3 {
		return BruteForce(px)
	}

	mid := len(px) / 2
	Qx := px[:mid]
	Rx := px[mid:]
	midX := Qx[len(Qx)-1].x

	Qy := []Point{}
	Ry := []Point{}
	for _, p := range py {
		if p.x <= midX {
			Qy = append(Qy, p)
		} else {
			Ry = append(Ry, p)
		}
	}

	p1, p2, distQ := ClosestPair(Qx, Qy)
	r1, r2, distR := ClosestPair(Rx, Ry)

	ss := math.Min(distQ, distR)
	bestPair := [2]Point{p1, p2}
	if distR < distQ {
		bestPair = [2]Point{r1, r2}
	}

	strip := []Point{}
	for _, p := range py {
		if math.Abs(p.x-midX) < ss {
			strip = append(strip, p)
		}
	}

	for i := 0; i < len(strip); i++ {
		for j := i + 1; j < len(strip) && strip[j].y-strip[i].y < ss; j++ {
			dist := Distance(strip[i], strip[j])
			if dist < ss {
				ss = dist
				bestPair = [2]Point{strip[i], strip[j]}
			}
		}
	}

	return bestPair[0], bestPair[1], ss
}

func ParseInput(scanner *bufio.Scanner) ([]Point, bool) {
	if !scanner.Scan() {
		return nil, false
	}

	n, _ := strconv.Atoi(scanner.Text())
	if n == 0 {
		return nil, false
	}

	points := make([]Point, n)
	for i := 0; i < n; i++ {
		if scanner.Scan() {
			line := scanner.Text()
			coords := strings.Fields(line)
			x, _ := strconv.ParseFloat(coords[0], 64)
			y, _ := strconv.ParseFloat(coords[1], 64)
			points[i] = Point{x, y}
		}
	}

	return points, true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	for {
		points, more := ParseInput(scanner)
		if !more {
			break
		}

		px := make([]Point, len(points))
		py := make([]Point, len(points))
		copy(px, points)
		copy(py, points)

		sort.Slice(px, func(i, j int) bool { return px[i].x < px[j].x })
		sort.Slice(py, func(i, j int) bool { return py[i].y < py[j].y })

		p1, p2, _ := ClosestPair(px, py)

		fmt.Printf("%.2f %.2f %.2f %.2f\n", p1.x, p1.y, p2.x, p2.y)
	}
}
